src/main/kotlin/snyk/common/lsp/SnykLanguageClient.kt

```diff
@@ -86,12 +87,62 @@ class SnykLanguageClient :
     private val progressEndMsgCache: Cache<String, WorkDoneProgressEnd> =
         Caffeine.newBuilder().expireAfterWrite(10, TimeUnit.SECONDS).build()
 
+
     override fun telemetryEvent(`object`: Any?) {
         // do nothing
     }
 
-    override fun publishDiagnostics(diagnostics: PublishDiagnosticsParams?) {
-        // do nothing for now
+    override fun publishDiagnostics(diagnosticsParams: PublishDiagnosticsParams?) {
+        if (diagnosticsParams == null) {
+            return
+        }
+
+        updateCache(diagnosticsParams)
+    }
+
+    private fun updateCache(diagnosticsParams: PublishDiagnosticsParams) {
+        val filePath = diagnosticsParams.uri
+
+        try {
+            getScanPublishersFor(filePath.toVirtualFile().path).forEach { (project, scanPublisher) ->
+                val snykFile = SnykFile(project, filePath.toVirtualFile())
+                val firstElement = diagnosticsParams.diagnostics.firstOrNull()
+
+                //If the diagnostics for the file is empty, clear the cache.
+                if (firstElement == null) {
+                    scanPublisher.onPublishDiagnostics("code", snykFile, emptyList())
+                    scanPublisher.onPublishDiagnostics("oss", snykFile, emptyList())
+                    return
+                }
+
+                // We always send PublishDiagnostics for one product. So if there is an item in the array, we can deduct the product.
+                val product = firstElement.source
```

Comment:
There is an interesting edge case here - if an error occurs, publishDiagnostics is called with URI = folder and one diagnostic that has `source` set to `Snyk Error`. In this case, we don't have a product, but should add it to the problems. I'm pretty sure that's possible, but some custom code.

Maybe have a look here: https://www.perplexity.ai/search/i-m-an-intellij-plugin-develop-O6QWPTa3RM6.G5SJT5pQpA

Reply:
Will handle this in a followup ticket.